lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -27,6 +27,11 @@ namespace {
 
 using namespace clang;
 
+// FIXME !!! Ensure that the implementation functions (all static functions
+// below) never call the public ASTStructuralEquivalence::IsEquivalent()
+// functions, because that will wreak havoc the internal state (DeclsToCheck
+// and Visited members) and could cause faulty behaviour or even infinite loop.
+
```

Comment:
Maybe not FIXME but NOTE or WARNING ?

Reply:
Is it not important to call only the `isStructurallyEquivalent(StructuralEquivalenceContext &, Decl*, Decl*)`, not the ones with more concrete `Decl` subtypes?